### ISSUE TITLE
?Error in simulate_sgd_update for MetaMatchingNetwork

### ISSUE THREAD

**Issue by luxius-luminus** [microsoft/AutoML#14]
Hello, thanks for your contribution!
I have found something confusing in the code. It would be great if you could help with these:

In
> lib/models/MetaMatchingNetwork.py

student weights are updated using SGD in line 19-25. 
But in  line 43-44, it writes
>def simulate_sgd_update(self, w, g, optimizer):
>return g * optimizer.param_groups[-1]['lr'] + w

which simulates the updated student weights via **gradient ascent** in order to take part in the calculation of teacher weights.

It turns out that updating student weights in "simulate_sgd_update" should be in accordance with updating student weights via SGD as in line 23. But in practice gradient **ascent** is used in one case while gradient **descent** is used in the other.

P.S. It also a little bit confusing to simulate the updated student weights right before you actually update them. Why not just take the updated weights from the student model? That's the **real** updated weights afterall.

**Comment by Z7zuqer** [comment]
Hi,

Thanks for your interest in our project!

* We define function *simulate_sgd_update* to simulate gradients update of student weights. [Real SGD updates](https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD) utilize momentum and weight_decay to update student weights. In the updates of MetaMN, we only focus on the updates of the Meta Matchnig Networks, ignoring the updates of student weights. So we only need an approximation of updated student weights to see whether updated student weights are better than before to evaluate the performance of MetaMN. Both gradient ascent and descent are all allowed here.

```
# SGD update in Pytorch Official Implemention
if weight_decay != 0:
    d_p = d_p.add(p, alpha=weight_decay)
if momentum != 0:
    param_state = self.state[p]
    if 'momentum_buffer' not in param_state:
        buf = param_state['momentum_buffer'] = torch.clone(d_p).detach()
    else:
        buf = param_state['momentum_buffer']
        buf.mul_(momentum).add_(d_p, alpha=1 - dampening)
    if nesterov:
        d_p = d_p.add(buf, alpha=momentum)
    else:
        d_p = buf

```
* In the procedure of updating MetaMN, only weights of meta matching networks should be updated and student weights should be fixed. To make sure student weights are well stored in the process of updating MetaMN, we don't use optimizer.step() to directly update the student weights. This is to prevent `momentum_buffer` in SGD optimizer to be affected(If 'momentum_buffer' is affected, student weights may not correctly reach the global minimum). So here we define the function *simulate_sgd_update* to simulate the updates of student weights.

Best,
Hao.

**Comment by lvxjl** [comment]
Hello Hao, thanks for you quick response!
I agree with you that it is sufficient to use an approximation of the updated student weights to update MetaMN.

However, in your response it says
>  So we only need ... whether **updated student weights** are better than before to **evaluate the performance of MetaMN**. 
Both gradient ascent and descent are all allowed here.

As the gradient is derived w.r.t the CE loss, updating weights by **gradient descent** drives the model towards a lower loss, while using **gradient ascent** moves it towards a higher.

But in the paper (and apparently), it says
> The underlying reason is that if one prioritized path αˆ∗k is complementary to the current subnetwork α, 
> then **the updated subnetwork with the weight w(t+1) α** can achieve **a lower loss on the validation data**

It turns out to be quite counterintuitive to see that using gradient ascent is OK.
And I am not sure if I fully understand how the performance of MetaMN depends on the weights of student network.
It would be great if you could provide more details, thank you.
